src/main/java/org/lantern/event/PublicIpAndTokenTracker.java

```diff
@@ -10,58 +11,52 @@
 import com.google.inject.Singleton;
 
 /**
- * This class centralizes tracking of the combination of a valid refresh 
- * token and a valid proxy. Several services rely on that combination,
- * including the XMPP server login and connecting to the friends API.
+ * This class centralizes tracking of the combination of a valid refresh token
+ * and a valid proxy. Several services rely on that combination, including the
+ * XMPP server login and connecting to the friends API.
  */
 @Singleton
 public class PublicIpAndTokenTracker {
 
     private final Logger log = LoggerFactory.getLogger(getClass());
-    
-    private boolean hasRefresh = false;
 
-    private boolean hasPublicIp = false;
+    private volatile boolean hasPublicIp = false;
+    private volatile String refresh;
 
-    private String refresh;
-    
     @Inject
     public PublicIpAndTokenTracker(final Model model) {
         log.debug("Creating tracker");
         Events.register(this);
         this.refresh = model.getSettings().getRefreshToken();
-        if (StringUtils.isNotBlank(refresh)) {
-            this.hasRefresh = true;
-        }
-    }
-    
-    /**
-     * Resets the state of the tracker (useful when reinitializing system).
-     */
-    public void reset() {
-        this.hasPublicIp = false;
     }
     
     @Subscribe
-    public void onRefreshToken(final RefreshTokenEvent refreshEvent) {
-        log.debug("Got refresh token -- loading friends");
-        this.refresh = refreshEvent.getRefreshToken();
-        synchronized (this) {
-            this.hasRefresh = true;
-            if (this.hasPublicIp) {
-                Events.asyncEventBus().post(new PublicIpAndTokenEvent(this.refresh));
-            }
+    synchronized public void onPublicIp(final PublicIpEvent pce) {
+        log.debug("Got public IP");
+        this.hasPublicIp = true;
+        if (StringUtils.isNotBlank(refresh)) {
+            log.debug("Have both public IP and refresh token, posting event");
+            Events.asyncEventBus()
+                    .post(new PublicIpAndTokenEvent(this.refresh));
         }
     }
-    
+
     @Subscribe
-    public void onPublicIp(final PublicIpEvent pce) {
-        synchronized (this) {
-            this.hasPublicIp = true;
-            if (this.hasRefresh) {
-                log.debug("Posting event!!");
-                Events.asyncEventBus().post(new PublicIpAndTokenEvent(this.refresh));
-            }
+    synchronized public void onConnectivityChanged(ConnectivityChangedEvent cce) {
```

Comment:
> So, PublicIpInfoHandler does the public ip check in response to a ProxyConnectionEvent, which happens as soon as the local proxy has started.

OK, so this is a related problem. With the new code on devel, there will only ever be a ProxyConnectionEvent once at startup. That means from that point on, we'll use that same public IP even if we change networks. It also means we'll never again generate a PublicIpEvent, which leads to other problems. So there's a bug with not learning about new IPs if we change networks for sure.

With the code on master, this doesn't happen, and reset() also always gets called before DefaultProxyTracker.init(), which means that reset always gets called before a PublicIpEvent is generated (since DefaultProxyTracker generates that event via generating a ProxyConnectionEvent). 

So it looks like both of these slipped through the cracks of the #2103 PR. In theory we could probably just make PublicIpInfoHandler actually do the new check on a ConnectivityChangedEvent for becoming connected instead of on ProxyConnectionEvent.

In terms of PublicIpAndTokenTracker the problem is now that it will never get a public IP after that first run. So when we get a ConnectivityChangedEvent for disconnecting, it will set hasPublicIp to false correctly, but PublicIpAndTokenTracker will never again get a PublicIpEvent because those are only generated in response to connections to the proxy. So the PublicIpAndTokenEvent will never get generated, and the friends screen or anything else waiting on that will hang.

I *think* just looking up the public IP in response to ConnectivityChanged might do it, but we have to somehow be sure the GetModeProxy has in fact started and is able to do the lookup.